### PR TITLE
tests: arch: arm: Fix thread swap test clobbering

### LINKED ISSUE #50709: tests: arch: arm: arm_thread_swap fails on stm32g0 or stm32l0
**Describe the bug**
running the tests/arch/arm/arm_thread_swap  on the   nucleo_g071rb or nucleo_l073rz will fail  (cortex M0+)

Please also mention any information which could help others to understand
the problem you're facing:
 In the tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c  the hard fault occurs after  
line 326  `z_move_thread_to_end_of_prio_q(_current);`
in  
`#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)`

**To Reproduce**
1. west build -p auto -b nucleo_g071rb tests/arch/arm/arm_thread_swap
2. west flash
3. See error

**Logs and console output**
```
*** Booting Zephyr OS build v3.2.0-rc2-233-g88705be89b8d  ***
Running TESTSUITE arm_thread_swap
===================================================================
START - test_arm_syscalls
 PASS - test_arm_syscalls in 0.001 seconds
===================================================================
START - test_arm_thread_swap
E: ***** HARD FAULT *****
E: r0/a1:  0x00000000  r1/a2:  0x20000504  r2/a3:  0x200001b8
E: r3/a4:  0x20000718 r12/ip:  0x89abcdef r14/lr:  0x08007203
E:  xpsr:  0x41000000
E: Faulting instruction address (r15/pc): 0x080009d8
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x200001b8 (unknown)
E: Halting system
```

**Environment (please complete the following information):**
 - OS: ubuntu
 - Toolchain : Zephyr SDK, 0.15.0
 - Commit SHA 88705be89b8d03de3df256bf5ade7a3b24a22866

**Additional context**
looks like https://github.com/zephyrproject-rtos/zephyr/issues/50146

### PR DESCRIPTION
**Pull request by microbuilder**
The thread swap test was making use of a register to point to `ztest_thread_callee_saved_regs_container`, which was being clobbered in the inline assembly code, causing a hard-fault in certain cases.

This commit adds a clobber list of the registers used, causing the compiler to assign `ztest_thread_callee_saved_regs_container` to an appropriate (unclobbered) register.

It also addresses an issue in the v6-m and v8-m.baseline inline assembler where only 4 of the 8 callee-saved registers were being set from `ztest_thread_callee_saved_regs_container`, due to a double use of `ldmia` on the same four registers. Since v6-m devices can't operate outside r0-r7 on most instructions, the upper 4 callee-saved registers are set first with `ldmia`, before setting the lower 4.

Fixes: #50709

Signed-off-by: Kevin Townsend <kevin.townsend@linaro.org>

### CODE REVIEW

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c [hunk 1]
```diff
@@ -344,16 +344,19 @@ static void alt_thread_entry(void)
 		"push {r0, r1};\n\t"
 		"mov r1, r7;\n\t"
 		"mov r0, %0;\n\t"
-		"ldmia r0!, {r4-r7};\n\t"
-		"ldmia r0!, {r4-r7};\n\t"
+		"ldr r4, [r0, #16];\n\t"
```
Comment (gromero) [inline tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:347]: How about keeping the 2 `ldmia` instructions instead of mixing `ldr` and `ldmia` here, like:

```
        "mov r0, %0;"
        "add r0, #16;"
        "ldmia r0!, {r4-r7};"
        "mov r8, r4;"
        "mov r9, r5;"
        "mov r10, r6;"
        "mov r11, r7;"
        "mov r0, %0;"
        "ldmia r0!, {r4-r7};"
```
? 

I think `load_callee_saved_regs` in line 64 should be fixed likewise -- for properly loading `r4-r7`?

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c [hunk 2]
```diff
@@ -344,16 +344,19 @@ static void alt_thread_entry(void)
...
+		"ldr r4, [r0, #24];\n\t"
+		"mov r10, r4;\n\t"
+		"ldr r4, [r0, #28];\n\t"
+		"mov r11, r4;\n\t"
+		"ldmia r0!, {r4-r7};\n\t"
 		"mov r7, r1;\n\t"
 		"pop {r0, r1};\n\t"
-		:	: "r" (&ztest_thread_callee_saved_regs_container)
-		: "memory"
+		: : "r" (&ztest_thread_callee_saved_regs_container)
```
Comment (gromero) [inline tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:358]: Could you put the output, input, and clobber lists in separate lines, like they appear in GCC manual:

```
: /* no output */
: "r" (&ztest_thread_callee_saved_regs_container)
: "memory", "r4", "r5", "r0", "r1", "r8", "r9", "r10", "r11"
```
? 

While you're at it, could you please add spaces between regs in the first `push` instruction `"memory", "r4", "r5", "r0", "r1", "r8", "r9", "r10", "r11"`?

**Review by gromero — CHANGES_REQUESTED** [review]
@microbuilder Hi Kevin. The changes look good. I just have a comment on using `ldr` and an additional fix in `load_callee_saved_regs` needed, but otherwise I agree the clobbering of  `r5` is the culprit for the fault issue in question. Feel free to wait for @FRASTM review before sending a v2. Cheers.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 1 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:347]: Thanks for the tips. Debugging, this causes a hardfault, though, because of:

```c
       "mov r0, %0;"
        "ldmia r0!, {r4-r7};"
```

GCC is resolving `%0` to `r7`, which we clobber in the first `"ldmia r0!, {r4-r7};"` (which sets `r7` to `0`), so the second `ldmia` call listed above is trying to increment `r0 = 0`, the same problem we had in the original issue.

I'll have to do some further debugging in the morning.

Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:347]: Also ... adding `r7` to the clobber list causes a compilation error and CI failure as well, in some situations, since this register has special meaning in the [Thumb Procedure Call Standard](https://developer.arm.com/documentation/dui0041/c/Thumb-Procedure-Call-Standard/TPCS-definition/TPCS-register-names?lang=en).

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 2 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c]: Done

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 1 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:347]: @gromero This should work. Rather than copying `%0` a second time (which has been clobbered as `r6` or `r7`), just decrement the pointer in `r0` by `#32`:

```
"mov r1, r7;\n\t"
"mov r0, %0;\n\t"
"add r0, #16;\n\t"
"ldmia r0!, {r4-r7};\n\t"
"mov r8, r4;\n\t"
"mov r9, r5;\n\t"
"mov r10, r6;\n\t"
"mov r11, r7;\n\t"
"sub r0, #32;\n\t"
"ldmia r0!, {r4-r7};\n\t"
```

**Comment by microbuilder** [comment]
@carlocaione @stephanosio Mind taking a second look?

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c [hunk 3]
```diff
@@ -68,25 +68,31 @@ static void load_callee_saved_regs(const _callee_saved_t *regs)
...
+		"mov r9, r5;\n\t"
+		"mov r10, r6;\n\t"
+		"mov r11, r7;\n\t"
+		"sub r0, #32;\n\t"
 		"ldmia r0!, {r4-r7};\n\t"
 		"mov r8, r4;\n\t"
 		"mov r9, r5;\n\t"
 		"mov r10, r6;\n\t"
 		"mov r11, r7;\n\t"
 		"mov r7, r1;\n\t"
```
Comment (stephanosio) [inline tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:83]: The second `ldmia` is loading `r4-r7` (i.e. `v1-v4`), so why are moving `r4-r7` to `r8-r11` here?

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c [hunk 4]
```diff
@@ -68,25 +68,31 @@ static void load_callee_saved_regs(const _callee_saved_t *regs)
...
-		:
+		: /* no output */
 		: "r" (regs)
-		: "memory"
+		: "memory", "r1", "r0", "r8", "r9", "r10", "r11"
```
Comment (stephanosio) [inline tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:86]: the purpose of this function is to load the content of `regs` into `v1-v8` (i.e. `r4-r11`), so does it really make sense to put these into the clobber list?

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c [hunk 5]
```diff
@@ -334,7 +340,7 @@ static void alt_thread_entry(void)
 	 */
 #if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
 	__asm__ volatile (
-		"push {r4,r5,r6,r7};\n\t"
+		"push {r4, r5, r6, r7};\n\t"
```
Comment (stephanosio) [inline tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:343]: Nit: can be simplified
```suggestion
		"push {r4-r7};\n\t"
```

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c [hunk 6]
```diff
@@ -344,16 +350,19 @@ static void alt_thread_entry(void)
...
 		"mov r11, r7;\n\t"
+		"sub r0, #32;\n\t"
+		"ldmia r0!, {r4-r7};\n\t"
 		"mov r7, r1;\n\t"
 		"pop {r0, r1};\n\t"
-		:	: "r" (&ztest_thread_callee_saved_regs_container)
-		: "memory"
+		: /* no output */
+		: "r" (&ztest_thread_callee_saved_regs_container)
+		: "memory", "r4", "r5", "r0", "r1", "r8", "r9", "r10", "r11"
```
Comment (stephanosio) [inline tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:365]: same deal here

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 3 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:83]: The first one retrieves the upper 4 values into r4-7 (input[4]), then pushes them into r8-11, since `ldmia` can only operate on r0-7 on the M0, etc. The second one moves the pointer back to input[0], then places those four words into r4-7, where they belong.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 4 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:86]: We only need r1 and r0 here, you're right.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 5 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:343]: ack

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 6 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:365]: If we don't have `r4`, `r5`, etc., in the clobber list, GCC will select `r4` for `%0`, which is what was causing the original problem in this code. We need these values to tell GCC which variables it can safely use for `%0`, which is only `r6` and `r7`. If we remove these, we'll end up in a hard fault.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 3 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:83]: Ugg, sorry. I missed your point here. You're right, those lines needs to be removed.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c [hunk 7]
```diff
@@ -68,25 +68,27 @@ static void load_callee_saved_regs(const _callee_saved_t *regs)
...
 		"mov r10, r6;\n\t"
 		"mov r11, r7;\n\t"
+		"sub r0, #32;\n\t"
+		"ldmia r0!, {r4-r7};\n\t"
 		"mov r7, r1;\n\t"
-		:
+		: /* no output */
 		: "r" (regs)
-		: "memory"
+		: "memory", "r1", "r0"
```
Comment (carlocaione) [inline tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: can you expand why it's not needed to have here `r8 / r9 / ...` as well? We are clobbering those AFAICT

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 6 above)
Comment (stephanosio) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:361]: well, this is a hairy situation ... it is playing a dangerous game with the compiler. IMHO, this whole thing needs to be written entirely in asm to guarantee correct behaviour, but that is beyond the scope of this PR ...

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 7 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: The clobber list is there for `%0`, which only occurs once early on, moving it to `r0`.  By the time anything happens to `r4-11`, we no longer have to worry about where `%0` was, since it's already been copied into `r0`, which is on the list (although normally GCC would only select r4-11 anyway).

Comment (stephanosio) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: the purpose of this function is to load `r4-r11` from a struct; adding these regs to the clobber list would nullify what this function does (because the compiler will try to restore the content of `r4-r11` "clobbered" in this inline asm block).

yes, it is a very dangerous thing to do ... but, with how this test is currently implemented, there is not much we can do.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 6 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:361]: I agree, the entire function should be written in assembly, but I'd like to at least fix the hard fault here, which this initial rework does.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 7 above)
Comment (carlocaione) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: right, but that really depends on the order of instructions and on the code itself, we have to remember ourselves to update the list if we modify that code.

Formally (IMO) we should list all the clobbered registers (after all, that is what the clobber list is for) and that includes `r8 / r9 / ...` as well.

Just a nit, I know.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 6 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:361]: I don't even understand why we're bothering with `ztest_thread_callee_saved_regs_container` if we're just zero'ing the values anyway. There are easier and cleaner solutions for that. But cleaning all of this up should likely be a new PR outside the current bug fix.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 7 above)
Comment (carlocaione) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: > because the compiler will try to restore the content of r4-r11 "clobbered" in this inline asm block

@stephanosio 

now I'm confused. I'm still scratching my head over the clobber list but according to the documentation `When the compiler selects which registers to use to represent input and output operands, it does not use any of the clobbered registers. As a result, clobbered registers are available for any use in the assembler code. `

Is GCC really supposed to restore the content of the clobbered registers?

Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: That's what I understood ... the clobber list is there to say what can safely be used for `%0`, etc. I'm not sure what to put here, though, since we're effectively using all of the general purpose registers that are available (`r4-11`). We're doing the same in the second instance, where at various points are are changing `r6` and `r7`, but those are the only registers we CAN use for `%0`.  

It seems like we have to have some awareness of the code when defining the clobber list. I know `r6` and `r7` are safe in the second instance, even though they are changed later, and we can't block EVERY register that gets changed. I don't like having to be aware of the code as well, but I don't see an alternative?

Comment (stephanosio) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: > Is GCC really supposed to restore the content of the clobbered registers?

If the compiler was already using one of the clobbered registers in the same context, it will have to save and restore the clobbered registers before and after, respectively, the inline asm block.

Comment (carlocaione) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: > If the compiler was already using one of the clobbered registers in the same context, it will have to save and restore the clobbered registers before and after, respectively, the inline asm block.

Ok, I got it now, thanks.

Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: It seems to me like we always need some awareness of our code when defining the clobber list, but it's a bit opaque to me as well.

For this code:

```c
	__asm__ volatile (
		"push {r4-r7};\n\t"
		"mov r4, r8;\n\t"
		"mov r5, r9;\n\t"
		"push {r4, r5};\n\t"
		"mov r4, r10;\n\t"
		"mov r5, r11;\n\t"
		"push {r4, r5};\n\t"
		"push {r0, r1};\n\t"
		"mov r1, r7;\n\t"
		"mov r0, %0;\n\t"
		"add r0, #16;\n\t"
		"ldmia r0!, {r4-r7};\n\t"
		"mov r8, r4;\n\t"
		"mov r9, r5;\n\t"
		"mov r10, r6;\n\t"
		"mov r11, r7;\n\t"
		"sub r0, #32;\n\t"
		"ldmia r0!, {r4-r7};\n\t"
		"mov r7, r1;\n\t"
		"pop {r0, r1};\n\t"
		: /* no output */
		: "r" (&ztest_thread_callee_saved_regs_container)
		: "memory", "r4", "r5", "r0", "r1", "r8", "r9", "r10", "r11"
	);
```

I need to restrict GCC from selecting `r4, r5, r0, r1` ... but with the `add` and `sub` changes, it seems safe to remove `r8-11` since those could be used in this case for the input/output registers, based on where they occur. I'd need to change my clobber list if `%0` occurred later, or was used again as it was in an earlier version of this code.

I may be missing some other nuance of the clobber list, though, in this case where we have inline ASM mixed with C code, which complicates things.

@stephanosio Would `: "memory", "r4", "r5", "r0", "r1"` be correct here in your opinion? Normally GCC wouldn't select `r0-3` anyway, but it seems prudent to put it just in case since we are using those before `%0`.

Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: Still trying to wrap my head around this.

If registers hold temporary values or are modified, it seems like they should be on the clobber list. See armclang documentation for another point of view: https://developer.arm.com/documentation/100067/0610/armclang-Inline-Assembler/Inline-assembly-statements-within-a-function/Clobber-list

> "if a register holds a temporary value, include it in the clobber list"

> "Each string is the name of a register that the assembly code potentially modifies, but for which the final value is not important."

But we're changing every register other than `r2` and `r3` here, which have their own reserved meaning.

Not sure I have the same conclusion as you, @stephanosio, of not including `r4-11` here, but what do you feel is the right list in this case to be sure I understand what you're proposing?

Excluding `r4` and `r5` from the list here will cause a hard fault, though, since we use those earlier in a temporary capacity. The problem is we use every register in a temporary/scratch manner, meaning it should be on the list, BUT registers 4-11 also have 'final values' that are important outside the function ... so in theory they should be excluded, but in practice we can't in this case.

Basically, this code is awkward and I don't think there's a right way to define the clobber list, other than being aware of the contents of the code so excluding everything modified before the input value (r0, r1, r4, r5).

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 4 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c]: This first case seems clear, it should only have r0 + r1, which I've updated. The discussion is around the much more complicated second case.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 7 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: I feel like the right list here is: `: "memory", "r4", "r5", "r0", "r1"` but let me know if there's no agreement on that.

**Comment by microbuilder** [comment]
The uneven `push` and `pop` functions here are also problematic in the second case.

From: https://gcc.gnu.org/onlinedocs/gcc-12.2.0/gcc/Extended-Asm.html#Clobbers-and-Scratch-Registers-1

> "the compiler requires the value of the stack pointer to be the same after an asm statement as it was on entry to the statement"

We need to balance the `pop` calls with the `push` calls, I believe.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 7 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: Was just discussing that elsewhere. We do need r4/r5 to be part for now, though, to avoid the hard fault.

However this code has a number of much larger issues. Are you objectionable to merging the current changes in as a solution to #50709, and I can do a rework of this code in a new PR?

**Comment by stephanosio** [comment]
> The uneven `push` and `pop` functions here are also problematic in the second case.

yea ... the whole thing really needs to be re-written entirely in asm.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 7 above)
Comment (stephanosio) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: No objection to the current solution since it does "fix" the issue and this needs to be reworked anyway.

Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: If you don't mind approving, I'll figure out a better solution for the larger problems with this code.

**Review by stephanosio — DISMISSED** [review]

**Comment by microbuilder** [comment]
> The uneven `push` and `pop` functions here are also problematic in the second case.
> 
> From: https://gcc.gnu.org/onlinedocs/gcc-12.2.0/gcc/Extended-Asm.html#Clobbers-and-Scratch-Registers-1
> 
> > "the compiler requires the value of the stack pointer to be the same after an asm statement as it was on entry to the statement"
> 
> We need to balance the `pop` calls with the `push` calls, I believe.

My mistake, we do `pop` those a bit later in the code in a new block, so it isn't unbalanced:

```c
#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
	__asm__ volatile (
		"pop {r4, r5, r6, r7};\n\t"
		"mov r8, r4;\n\t"
		"mov r9, r5;\n\t"
		"mov r10, r6;\n\t"
		"mov r11, r7;\n\t"
		"pop {r4, r5, r6, r7};\n\t"
		: : :
	);
#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
	__asm__ volatile (
		"pop {v1-v8};\n\t"
		: : :
	);
#endif
```

Although having this in two blocks seems to go against the spirit of what GCC is saying, that a block should enter/exit with the same SP, so I can try to bring those together in the rework.

**Review by carlocaione — DISMISSED** [review]

**Comment by gromero** [comment]
> Although having this in two blocks seems to go against the spirit of what GCC is saying, that a block should enter/exit with the same SP, so I can try to bring those together in the rework.

@microbuilder I agree. For that I think we can just use an alternative stack to stash `r4-r11` instead of using the function's stack. But I think that's out-of-scope for the PR, which is primarily related to the clobber issue.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 7 above)
Comment (gromero) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:82]: @microbuilder PTAL at https://github.com/zephyrproject-rtos/zephyr/pull/51366#discussion_r999631890 about the clobber list, I agree with @stephanosio and I propose a tweak to accommodate it. HTH.

**Review by gromero — CHANGES_REQUESTED** [review]
Did another review round considering @stephanosio comments on the clobber list here https://github.com/zephyrproject-rtos/zephyr/pull/51366#discussion_r999631890 @microbuilder PTAL.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 6 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:361]: @gromero Thanks for taking the time to sort some of this out and make a proposal. Happy to test this on HW, and update the PR before we merge.

**Comment by microbuilder** [comment]
Marking as `DNM` pending testing of suggestions from @gromero

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 6 above)
Comment (gromero) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:361]: @microbuilder np. Btw, with the "ldmia" form without `!` doesn't work, please revert to your previous code, i.e. change the code pasted above the following:

```
diff --git a/ldmia.c b/ldmia.c
index 81d8fd7..255d90e 100644
--- a/ldmia.c
+++ b/ldmia.c
@@ -66,13 +66,13 @@ int main()
                  */
                 "mov r0, %0;" // since r0 is in the clobber list r0 is not allocated to %0, so r0 != register(%0)
                 "add r0, #16;"
-                "ldmia r0, {r4-r7};"
+                "ldmia r0!, {r4-r7};"
                 "mov r8, r4;"
                 "mov r9, r5;"
                 "mov r10, r6;"
                 "mov r11, r7;"
-                "sub r0, #16;"
-                "ldmia r0, {r4-r7};"
+                "sub r0, #32;"
+                "ldmia r0!, {r4-r7};"
 
                 /*
                  * Restore r0 and r7
```

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c [hunk 8]
```diff
@@ -334,26 +336,44 @@ static void alt_thread_entry(void)
...
-		"pop {r0, r1};\n\t"
-		:	: "r" (&ztest_thread_callee_saved_regs_container)
-		: "memory"
+		/* Stash r4-r11 in stack, they will be restored much later in
+		 * another inline asm -- that should be reworked since stack
+		 * must be balanced when we leave any inline asm. We could
+		 * use simply an alternative stack for storing them instead
+		 * of the function's stack.
+		 */
+		"push {r4-r7};"
```
Comment (gromero) [inline tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:345]: @microbuilder Sorry, I missed that, but you need to add `\n\t` in all these lines. I tend to not use them, but to be consist with the rest of the code in the test I think it's good to add them likewise here.

**Review by gromero — CHANGES_REQUESTED** [review]
@microbuilder Just a nit on missing `\n\t`. See comment inline.

**Review by stephanosio — DISMISSED** [review]

**Review by gromero — APPROVED** [review]
@microbuilder LGTM! @stephanosio @carlocaione @carlescufi I have not further comments on it. Thanks.

File: tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c (hunk 6 above)
Comment (microbuilder) [reply tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c]: @gromero I noticed that stepping through the code, and looking at the instruction in the armasm PDF for Thumb. Already pushed a fix out, current code should be good for final review.

**Review by stephanosio — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]
LGTM. This was a very instructive PR, thank you for this :)
